pkg/remote/write.go

```diff
@@ -0,0 +1,232 @@
+package remote
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sort"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage/remote"
+
+	mcontext "github.com/grafana/prometheus-pulsar-remote-write/pkg/context"
+	"github.com/grafana/prometheus-pulsar-remote-write/pkg/pulsar"
+	"github.com/grafana/prometheus-pulsar-remote-write/pkg/version"
+)
+
+type Write struct {
+	logger log.Logger
+
+	samplePerTenantID   map[string][]pulsar.ReceivedSample
+	deadlinePerTenantID map[string]time.Time
+
+	BatchSize     int
+	BatchMaxDelay time.Duration
+}
+
+type WriteOpts func(o *Write)
+
+func WithLogger(l log.Logger) WriteOpts {
+	return func(w *Write) {
+		w.logger = l
+	}
+}
+
+func NewWrite(opts ...WriteOpts) *Write {
+	w := &Write{
+		logger: log.NewNopLogger(),
+
+		samplePerTenantID:   make(map[string][]pulsar.ReceivedSample),
+		deadlinePerTenantID: make(map[string]time.Time),
+
+		BatchSize:     100,
+		BatchMaxDelay: 5 * time.Second,
+	}
+
+	for _, opt := range opts {
+		opt(w)
+	}
+
+	return w
+}
+
+type headerRoundtripper struct {
+	upstream http.RoundTripper
+}
+
+func (t *headerRoundtripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// if tenant ID is set, expose it in the header
+	tenantID := mcontext.TenantIDFromContext(req.Context())
+	if tenantID != "" {
+		req.Header.Set(mcontext.HTTPHeaderTenantID, tenantID)
+	}
+
+	// overwrite the user agent header
+	req.Header.Set("User-Agent", fmt.Sprintf("%s/%s", version.AppName(), version.AppVersion()))
+
+	return t.upstream.RoundTrip(req)
+}
+
+type ClientConfig = remote.ClientConfig
+
+func NewWriteClient(conf *ClientConfig) (remote.WriteClient, error) {
+	clientInterface, err := remote.NewWriteClient("remote-write", conf)
+	if err != nil {
+		return nil, err
+	}
+
+	client, ok := clientInterface.(*remote.Client)
+	if !ok {
+		return nil, fmt.Errorf("client of unexpected type %T", clientInterface)
+	}
+
+	// add custom rountripper to modify user-agent and tenant header if necessary
+	client.Client.Transport = &headerRoundtripper{upstream: client.Client.Transport}
+
+	return client, nil
+}
+
+func (w *Write) Run(ctx context.Context, sampleCh chan pulsar.ReceivedSample, client remote.WriteClient) error {
+	tick := time.NewTicker(100 * time.Millisecond)
+	defer tick.Stop()
+
+	// this is set when a retry able error happend
+	errRemoteWriteRetryable := false
```

Comment:
Where could this get set to `true`?

Reply:
Thanks that's something I have missed, also shows the need for testing :slightly_smiling_face: